### PR TITLE
Get eth1 data using eth1 block timestamp

### LINKED ISSUE #704: Use timestamp instead of block height in eth1 voting to allow for simpler caching
https://github.com/ethereum/eth2.0-specs/pull/1553

### PR DESCRIPTION
**Pull request by twoeths**
## Goal
+ Close #704
+ The main thing is to implement an eth1 block cache in order to get eth1 data

### CODE REVIEW

File: packages/lodestar/src/eth1/impl/ethers.ts [hunk 1]
```diff
@@ -63,19 +70,53 @@ export class EthersEth1Notifier extends (EventEmitter as { new(): Eth1EventEmitt
...
     this.emit("block", block);
+    let requestedBlockNumber = this.blocksCache.requestNewBlock(block);
+    if (requestedBlockNumber) {
+      let requestedBlock: Block;
+      let retry = 0;
+      while (!requestedBlock && retry < GET_ETH1_BLOCK_RETRY) {
+        requestedBlock = await this.getBlock(requestedBlockNumber);
+        requestedBlockNumber++;
+        retry++;
+      }
```
Comment (mpetrunic) [inline packages/lodestar/src/eth1/impl/ethers.ts:114]: We are doing this in bunch of places now. Can you create some utility function like:
```javascript
export async function retryable<T>(func: (...args) => T|null , retry = 3): T|null;
```
so we would have:
```javascript
const requestedBlock =  retryable(this.getBlock.bind(this, requestedBlockNumber), GET_ETH1_BLOCK_RETRY)
```

File: packages/lodestar/src/eth1/impl/eth1Vote.ts [hunk 2]
```diff
@@ -1,36 +1,14 @@
...
-    votesToConsider = await getEth1DataRange(
-      this.getEth1Data,
-      eth1Head,
-      config.params.ETH1_FOLLOW_DISTANCE,
-      2 * config.params.ETH1_FOLLOW_DISTANCE
-    );
-  }
+  this: IEth1Notifier, config: IBeaconConfig, state: BeaconState): Promise<Eth1Data> {
+  const periodStart = votingPeriodStartTime(config, state);
+  const promises: Promise<Eth1Data>[] = this.findBLocks(config, periodStart).map(block => this.getEth1Data(block));
```
Comment (mpetrunic) [inline packages/lodestar/src/eth1/impl/eth1Vote.ts]:
```suggestion
  const promises: Promise<Eth1Data>[] = this.findBlocks(config, periodStart).map(block => this.getEth1Data(block));
```

Comment (mpetrunic) [reply packages/lodestar/src/eth1/impl/eth1Vote.ts]: rename method implementation also

**Review by mpetrunic — COMMENTED** [review]
Haven't checked against spec yet

File: packages/lodestar/src/eth1/impl/eth1Vote.ts (hunk 2 above)
Comment (twoeths) [reply packages/lodestar/src/eth1/impl/eth1Vote.ts]: not sure if I understand this correctly. Do you mean to rename the `findBlocks` method?

Comment (mpetrunic) [reply packages/lodestar/src/eth1/impl/eth1Vote.ts]: notice lowercase L

Comment (twoeths) [reply packages/lodestar/src/eth1/impl/eth1Vote.ts]: I see it now lol

File: packages/lodestar/src/eth1/impl/ethers.ts (hunk 1 above)
Comment (twoeths) [reply packages/lodestar/src/eth1/impl/ethers.ts:114]: I have to increase `requestedBlockNumber` on every call so have to come up with something like this which is quite complex:
```
export async function retryable<T>(funcGen: Generator<(...args: any[]) => Promise<T|null>>, retry = 3): Promise<T|null> {
  let result: T;
  let count = 0;
  while(!result && count < retry) {
    const func = funcGen.next().value;
    result = await func();
    count++;
  }
  return result;
}
```
And the consumers have to implement a generator. Any simpler approach?

Comment (mpetrunic) [reply packages/lodestar/src/eth1/impl/ethers.ts:114]: I think you should support plain function as well and check inside `retryable` if argument is iterator or plain function

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**